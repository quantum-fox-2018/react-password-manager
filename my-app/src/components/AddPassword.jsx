import React, { Component } from 'react';
import {addPassword} from '../store/password/password.action'
import {getUser} from '../store/users/users.action'
import {connect} from 'react-redux'
import {bindActionCreators} from 'redux'

class AddPassword extends Component {
  constructor () {
    super ()
    this.state = {
      url: '',
      username:'',
      password: ''
    }
  }

  handleChange = (e) => {
    e.preventDefault()
    this.setState({
      [e.target.name] : e.target.value
    })
  }

  addPasswordButton = () => {
    let {history} = this.props
    let user = localStorage.getItem('username')
    let newPass = {
      url : this.state.url,
      username: this.state.username,
      password: this.state.password
    }
    this.props.addPassword(newPass, user)
    history.push('/home')
  }
  lowerCase () {
    return (this.state.password.match(/^(?=.*[a-z])/))
  }
  
  upperCase () {
    return (this.state.password.match(/^(?=.*[A-Z])/)) 
  }
  
  oneNum () {
    return (this.state.password.match(/^(?=.*[0-9])/))
  }
  
  minLength () {
    return (this.state.password.length > 6) ? true : false
  }

  specialChar () {
    return (this.state.password.match(/^(?=.*[_\W])/))
  }

  allValid() {
    return (this.minLength() && this.upperCase() && this.lowerCase() && this.oneNum() && this.specialChar())
  }
  render() {
    return (
      <div className="container">
        <h3>Add new url</h3>
        <form>
          <div className="form-group">
            <label>URL</label>
            <input type="text" className="form-control" name="url" placeholder="Enter url.."
            value={this.state.url}
            onChange={this.handleChange}
            />
          </div>
          <div className="form-group">
            <label>Username</label>
            <input type="text" className="form-control" name="username" placeholder="Enter username"
              value={this.state.username}
              onChange={this.handleChange}
            />
          </div>
          <div className="form-group">
            <label>Password</label>
            <input type="password" className="form-control" name="password" placeholder="Password"
              value={this.state.password}
              onChange={this.handleChange}
            />
          </div>
          <button type="button" className={"btn btn-primary "+ (this.allValid() ? "active" : "disabled")} onClick={this.addPasswordButton}>Submit</button>
        </form>
        <div style={{marginTop:'10px'}}>
          <h4>Password requirements: </h4>
            <ul className="text-left" style={{listStyle: "none"}}>
              <li className={"alert alert-dismissible " + (this.lowerCase() ? "alert-success" : "alert-danger")}> {this.lowerCase() ? '[ O ]': '[ X ]'} At least <strong>one lowercase</strong></li>
              <li className={"alert alert-dismissible " + (this.upperCase() ? "alert-success" : "alert-danger")}> {this.upperCase() ? '[ O ]': '[ X ]'} At least <strong>one capital</strong></li>
              <li className={"alert alert-dismissible " + (this.oneNum() ? "alert-success" : "alert-danger")}> {this.oneNum() ? '[ O]': '[ X ]'} At least <strong>one number</strong></li>
              <li className={"alert alert-dismissible " + (this.minLength() ? "alert-success" : "alert-danger")}> {this.minLength() ? '[ O ]': '[ X ]'} Minimum <strong>6 characters</strong></li>
              <li className={"alert alert-dismissible " + (this.specialChar() ? "alert-success" : "alert-danger")}> {this.specialChar() ? '[ O ]': '[ X ]'} At least <strong>Use one symbol</strong></li>
            </ul>
          </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => bindActionCreators({
  addPassword,
  getUser
}, dispatch)

const mapStateToProps = (state) => ({
  user: state.user
})

export default connect(mapStateToProps, mapDispatchToProps) (AddPassword);