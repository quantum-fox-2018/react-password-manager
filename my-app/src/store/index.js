import {createStore, applyMiddleware, combineReducers} from 'redux'
import thunk from 'redux-thunk'
import passwordReducers from './password/password.reducers'
import userReducers from './users/user.reducers'
import logger from 'redux-logger'


const reducers = combineReducers({
  passManager : passwordReducers,
  user: userReducers
})

const store = createStore(
  reducers,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
  applyMiddleware(thunk, logger)
)

export default store