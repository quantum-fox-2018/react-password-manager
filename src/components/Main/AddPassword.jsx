import React, { Component } from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { AddPassword as AddPasswordAction } from '../../store/password/password.action'
import PasswordValidation from './PasswordValidation'

class AddPassword extends Component {
  constructor() {
    super()
    this.state = {
      newUrl: '',
      newEmail: '',
      newPassword: '',
      upperCase: false,
      lowerCase: false,
      specialChar: false,
      number: false,
      minLength: false,
      isItValid: false
    }
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    }, () => {
      this.uCaseValidation()
      this.lCaseValidation()
      this.sCharValidation()
      this.numberValidation()
      this.lengthValidation()
    })
  }
  
  uCaseValidation = () => {
    let upperCase = /[A-Z]/g.test(this.state.newPassword)
    if (upperCase) {
      this.setState({ upperCase: true }, () => this.checkAllValidation())
    } else {
      this.setState({ upperCase: false })
    }
    // console.log('upperCase ', upperCase)
  }

  lCaseValidation = () => {
    let lowerCase = /[a-z]/g.test(this.state.newPassword)
    if (lowerCase) {
      this.setState({ lowerCase: true }, () => this.checkAllValidation())
    } else {
      this.setState({ lowerCase: false })
    }
    // console.log('lowerCase ', lowerCase)
  }

  sCharValidation = () => {
    // \W => nonword char
    let specialChar = /\W/g.test(this.state.newPassword)
    if (specialChar) {
      this.setState({ specialChar: true }, () => this.checkAllValidation())
    } else {
      this.setState({ specialChar: false })
    }
    // console.log('special char ', specialChar)
  }

  numberValidation = () => {
    // \d digit char
    let number =  /\d/g.test(this.state.newPassword)
    if (number) {
      this.setState({ number: true }, () => this.checkAllValidation())
    } else {
      this.setState({ number: false })
    }
    // console.log('number ', number)
  }

  lengthValidation = () => {
    if (this.state.newPassword.length < 6) {
      // console.log('length', false)
      this.setState({ minLength: false })
    } else {
      // console.log('length', true)
      this.setState({ minLength: true }, () => this.checkAllValidation())
    }
  }

  checkAllValidation = () => {
    let addButton = document.querySelector('#addPass')
    let upperCase = this.state.upperCase
    let lowerCase = this.state.lowerCase
    let specialChar = this.state.specialChar
    let number = this.state.number
    let minLength = this.state.minLength
    let url = this.state.newUrl
    let email = this.state.newEmail 

    if (upperCase && lowerCase && specialChar &&
    number && minLength) {
      // pas testing bikin error soalny addButton null pas render pertama kali
      if (addButton && url && email) {
        addButton.classList.remove('disabled')
      }
      this.setState({isItValid: true})
    } else {
      if (addButton && url && email) {
        addButton.classList.add('disabled')
      }
      this.setState({isItValid: false})
    }
  }

  resetState = () => {
    this.setState({
      ...this.state,
      newUrl: '',
      newEmail: '',
      newPassword: ''
    })
    // bikin error pas di testing karena getElement sebelum render
    let url = document.getElementById('newUrl')
    let email = document.getElementById('newEmail')
    let pass = document.getElementById('newPassword')
    if ( url && email && pass ) {
      url.value = ''
      email.value = ''
      pass.value = ''
    }
  }

  submitDataPass = (e) => {
    e.preventDefault()
    let userId = this.props.user.userId
    let newPass = {
      url: this.state.newUrl,
      email: this.state.newEmail,
      password: this.state.newPassword
    }
    // console.log('Add Data Pass ===>', newPass)
    this.props.AddPasswordAction(newPass, userId)
    this.resetState()
  }

  render() {
    let PassVal = ''
    if (!this.state.isItValid) {
      PassVal = (
        <PasswordValidation 
          className="col s12"
          upperCase={ this.state.upperCase } 
          lowerCase={ this.state.lowerCase }
          specialChar={ this.state.specialChar }
          number={ this.state.number }
          minLength={ this.state.minLength }
        />
      )
    }
                  
    return (
      <div className="container">
        <div className="card-panel">
          <div className="row">
            <form className="col s12" onSubmit={this.submitDataPass}>
              <h4 className="header2">Add New Password</h4>
              <div className="row">
                <div className="input-field col s12 m3">
                  <input id="newUrl" name="newUrl" type="text" 
                  className="validate" value={this.state.newUrl}
                  onChange={this.handleChange} required/>
                  <label htmlFor="newUrl">Url</label>
                </div>
                <div className="input-field col s12 m3">
                  <input id="newEmail" name="newEmail" type="text" 
                  className="validate" value={this.state.newEmail}
                  onChange={this.handleChange} required/>
                  <label htmlFor="newEmail">Email / Username</label>
                </div>
                <div className="input-field col s12 m3">
                  <input id="newPassword" name="newPassword" type="password" 
                  className="validate" value={this.state.newPassword}
                  onChange={this.handleChange} required/>
                  <label htmlFor="newPassword">Password</label>
                </div>
                <div className="input-field col s12 m3">
                  <button type="submit" className="btn teal waves-effect waves-light disabled" id="addPass">Add</button>
                </div>
              </div>
              {PassVal}
            </form>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  user: state.user,
  passwords: state.passwords
})

const mapDispatchToProps = (dispatch) => bindActionCreators({
  AddPasswordAction
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(AddPassword)