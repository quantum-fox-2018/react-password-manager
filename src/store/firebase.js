import firebase from 'firebase'

const config = {
  apiKey: "AIzaSyCJ0Hlgag0DOOKpxzUsGTaE4FAu0tIE_EE",
  authDomain: "react-password-manager-bef30.firebaseapp.com",
  databaseURL: "https://react-password-manager-bef30.firebaseio.com",
  projectId: "react-password-manager-bef30",
  storageBucket: "react-password-manager-bef30.appspot.com",
  messagingSenderId: "827996324966"
}
firebase.initializeApp(config)
export const Database = firebase.database()
export const User = firebase.auth()