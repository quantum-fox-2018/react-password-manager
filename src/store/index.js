import { createStore, applyMiddleware, combineReducers } from "redux"
import logger from 'redux-logger'
import usersReducers from './users/reducers'
import thunk from 'redux-thunk'
import passmanagerReducer from "./passmanager/reducers";

const reducers = combineReducers({
  users: usersReducers,
  passmanager: passmanagerReducer
})

let store = createStore(reducers, window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(), applyMiddleware(thunk, logger))

export default store