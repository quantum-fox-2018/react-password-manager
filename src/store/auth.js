import { observable } from 'mobx'
import swal from 'sweetalert2'
import { auth } from './firebase'


class Auth {
  userId = ''
  email = ''
  @observable statusAuth = false

  errorNotif(err) {
    swal({
      position: 'center',
      type: 'error',
      title: 'Check your password',
      text: err.message
    })
  }

  successNotif(info) {
    swal({
      position: 'center',
      type: 'success',
      title: info,
      showConfirmButton: false,
      timer: 1000
    })
  }

  signUp(email, password) {
    auth.createUserWithEmailAndPassword(email, password)
    .then(res => {
      this.email = email
      this.userId = res.uid
      this.statusAuth = true
      this.successNotif('Enjoy!')
    })
    .catch(err => {
      this.errorNotif(err)
    })
  }

  signIn(email, password) {
    auth.signInWithEmailAndPassword(email, password)
    .then(res => {
      this.email = email
      this.userId = res.uid
      this.statusAuth = true
      this.successNotif('Welcome Back!')
    })
    .catch(err => {
      this.errorNotif(err)
    })
  }

  signOut() {
    this.email = ''
    this.userId = ''
    auth.signOut()
    .then(res => {
      this.successNotif('See You!')
      this.statusAuth = false
    })
    .catch(err => {
      this.errorNotif(err)
    })
  }
}

export default new Auth()
