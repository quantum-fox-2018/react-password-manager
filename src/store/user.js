import { observable } from 'mobx'
import swal from 'sweetalert2'


import { db, auth } from './firebase'
import authData from './auth'

class PassManager {
  @observable listData = []
  valUrl = false
  valLength = false
  valNumber = false
  valUpper = false
  valLower = false
  valSpecial = false
  allVal = false

  checkUrl(url) {
    let regex = /[-a-zA-Z0-9@:%_+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_+.~#?&//=]*)?/gi
    this.valUrl = regex.test(url)
    return this.valUrl
  }

  checkLength(password) {
    this.valLength = (password.length >= 6)
    return this.valLength
  }

  checkNumber(password) {
    this.valNumber = (/\d/g.test(password))
    return this.valNumber
  }

  checkUpper(password) {
    this.valUpper = (/[A-Z]/g.test(password))
    return this.valUpper
  }

  checkLower(password) {
    this.valLower = (/[a-z]/g.test(password))
    return this.valLower
  }

  checkSpecial(password) {
    this.valSpecial = (/\W/g.test(password))
    return this.valSpecial
  }

  checkAllValidation() {
    if(this.valUrl && this.valLength && this.valNumber && this.valUpper && this.valLower && this.valSpecial) {
      this.allVal = true
      return this.allVal
    }
  }

  hashPassword(str) {
    let length = str.length
    let bullet = String.fromCharCode(0x2022)
    return bullet.repeat(length)
  }

  showPassword(userId, idData) {
    swal.mixin({
      input: 'password',
      confirmButtonText: 'Confirm',
      showCancelButton: true,
    }).queue([
      {
        title: 'Password',
        text: 'enter your password'
      }
    ]).then((result) => {
      if (!result.value) {
        swal({
          type: 'question',
        })
      } else if (result.value[0]) {
        auth.signInWithEmailAndPassword(authData.email, result.value[0])
        .then(res => {
          let data = this.listData.map(data => {
            if(data.id === idData) {
              return {...data, showPassword: !data.showPassword}
            } else {
              return {...data}
            }
          })
          this.listData = data  
        })
        .catch(err => {
          swal({
            type: 'error',
            title: 'Error',
            text: err.message
          })
        })
      }
    })
  }

  loadData(userId) {
    db.ref(`/user/${userId}`).once('value', (snap) => {
      const pass = snap.val()
      let arr = []
      for (const key in pass) {
        if (pass.hasOwnProperty(key)) {
          const el = pass[key]
          const update = {...el, id: key, showPassword: false}
          arr.push(update)
        }
      }
      this.listData = arr
    })
  }

  createData(userId, url, username, password) {
    db.ref(`/user/${userId}`).push({
      url, username, password, createdAt: Date.now(), updatedAt: null
    })
    .then(res => {
      swal({
        position: 'center',
        type: 'success',
        title: 'Your password has been saved',
        showConfirmButton: false,
        timer: 1500
      })
      this.loadData(userId)
    })
    .catch(err => {
      alert(err)
    })
  }

  deleteData(userId, idData) {
    swal({
      title: 'Are you sure?',
      text: "You won't be able to revert this!",
      type: 'warning',
      showCancelButton: true,
      confirmButtonColor: '#3085d6',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Yes, delete it!'
    }).then((result) => {
      if (result.value) {
        db.ref(`/user/${userId}`).child(idData).remove()
        this.loadData(userId)
        swal(
          'Deleted!',
          'Your file has been deleted.',
          'success'
        )
      }
    })
  }

  editPassword(userId, idData) {
    swal.mixin({
      input: 'password',
      confirmButtonText: 'Confirm',
      showCancelButton: true,
      inputAttributes: {
        'minlength': 6,
        'autocapitalize': 'off',
        'autocorrect': 'off'
      }
    }).queue([
      {
        title: 'Edit Password',
        text: 'enter your new password'
      }
    ]).then((result) => {
      if (result.value) {
        let data = this.listData.map(data => {
          if(data.id === idData) {
            let newData = {...data, password: result.value[0], updatedAt: Date.now()}
            db.ref(`/user/${userId}`).child(idData).update(newData)
            return newData
          } else {
            return {...data}
          }
        })
        this.listData = data
        swal({
          position: 'center',
          type: 'success',
          title: 'Your password has been edited',
          showConfirmButton: false,
          timer: 1500
        })
      } else {
        swal({
          type: 'question',
        })
      }
    })
  }
}

export default new PassManager()
