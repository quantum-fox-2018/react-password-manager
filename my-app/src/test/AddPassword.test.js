import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, {shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {Provider} from 'react-redux'
import {Route, Switch} from 'react-router-dom'
import store from '../store'
import LocalStorageMock from './localstorageMock'
import renderer from 'react-test-renderer'
import AddPassword from '../components/AddPassword'

global.localStorage = new LocalStorageMock();
Enzyme.configure({ adapter: new Adapter() });


describe('<AddPassword/>', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    const wrapper = mount(
    <Provider store={store}>
      <AddPassword/>
    </Provider>
    )
    ReactDOM.render(wrapper, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders without crashing snapshot', () => {
    const rendered = renderer.create(
    <Provider store={store}>
      <AddPassword/>
    </Provider>)
    expect(rendered).toMatchSnapshot()
  });
})