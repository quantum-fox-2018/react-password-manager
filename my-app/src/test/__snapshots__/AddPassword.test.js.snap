// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddPassword/> renders without crashing snapshot 1`] = `
<div
  className="container"
>
  <h3>
    Add new url
  </h3>
  <form>
    <div
      className="form-group"
    >
      <label>
        URL
      </label>
      <input
        className="form-control"
        name="url"
        onChange={[Function]}
        placeholder="Enter url.."
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label>
        Username
      </label>
      <input
        className="form-control"
        name="username"
        onChange={[Function]}
        placeholder="Enter username"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label>
        Password
      </label>
      <input
        className="form-control"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <button
      className="btn btn-primary disabled"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </form>
  <div
    style={
      Object {
        "marginTop": "10px",
      }
    }
  >
    <h4>
      Password requirements: 
    </h4>
    <ul
      className="text-left"
      style={
        Object {
          "listStyle": "none",
        }
      }
    >
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          one lowercase
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          one capital
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          one number
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         Minimum 
        <strong>
          6 characters
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          Use one symbol
        </strong>
      </li>
    </ul>
  </div>
</div>
`;
